src/chunked/encoder.rs

```diff
@@ -0,0 +1,185 @@
+    }
+    /// Encode an AsyncBufRead using "chunked" framing. This is used for streams
+    /// whose length is not known up front.
+    ///
+    /// # Format
+    ///
+    /// Each "chunk" uses the following encoding:
+    ///
+    /// ```txt
+    /// 1. {byte length of `data` as hex}\r\n
+    /// 2. {data}\r\n
+    /// ```
+    ///
+    /// A chunk stream is finalized by appending the following:
+    ///
+    /// ```txt
+    /// 1. 0\r\n
+    /// 2. {trailing header}\r\n (can be repeated)
+    /// 3. \r\n
+    /// ```
+    pub(crate) fn encode(
+        &mut self,
+        res: &mut Response,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        self.bytes_written = 0;
+        match self.state {
+            State::Streaming => self.encode_stream(res, cx, buf),
+            State::ReceiveTrailers => self.encode_trailers(res, cx, buf),
+            State::EncodeTrailers => self.encode_trailers(res, cx, buf),
+            State::EndOfStream => self.encode_eos(cx, buf),
+            State::Done => Poll::Ready(Ok(0)),
+        }
+    }
+
+    /// Stream out data using chunked encoding.
+    fn encode_stream(
+        &mut self,
+        mut res: &mut Response,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        // Get bytes from the underlying stream. If the stream is not ready yet,
+        // return the header bytes if we have any.
+        let src = match Pin::new(&mut res).poll_fill_buf(cx) {
+            Poll::Ready(Ok(n)) => n,
+            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+            Poll::Pending => match self.bytes_written {
+                0 => return Poll::Pending,
+                n => return Poll::Ready(Ok(n)),
+            },
+        };
+
+        // If the stream doesn't have any more bytes left to read we're done.
+        if src.len() == 0 {
+            // Write out the final empty chunk
+            let idx = self.bytes_written;
+            buf[idx] = b'0';
+            buf[idx + 1] = CR;
+            buf[idx + 2] = LF;
+            self.bytes_written += 3;
+
+            self.state = State::ReceiveTrailers;
+            return self.receive_trailers(res, cx, buf);
+        }
+
+        // Each chunk is prefixed with the length of the data in hex, then a
+        // CRLF, then the content, then another CRLF. Calculate how many bytes
+        // each part should be.
+        let buf_len = buf.len().checked_sub(self.bytes_written).unwrap_or(0);
+        let amt = src.len().min(buf_len);
+        // Calculate the max char count encoding the `len_prefix` statement
+        // as hex would take. This is done by rounding up `log16(amt + 1)`.
+        let hex_len = ((amt + 1) as f64).log(16.0).ceil() as usize;
+        let crlf_len = 2 * 2;
+        let buf_upper = buf_len.checked_sub(hex_len + crlf_len).unwrap_or(0);
+        let amt = amt.min(buf_upper);
+        let len_prefix = format!("{:X}", amt).into_bytes();
```

Comment:
No need to allocate using `into_bytes`, you can use `as_bytes`, to determine the length and to copy it over.